site/en/blog/renderingng-architecture/index.md

```diff
@@ -0,0 +1,508 @@
+---
+layout: 'layouts/blog-post.njk'
+title: 'Overview of the RenderingNG architecture'
+description: >
+  This post explains the components of the RenderingNG architecture,
+  and how the rendering pipeline flows through them.
+date: 2021-07-21
+authors:
+  - chrishtr
+---
+
+
+In a [previous post](/blog/renderingng/),
+I gave an overview of the RenderingNG architecture goals and key properties.
+This post will explain how its component pieces are set up,
+and how the rendering pipeline flows through them.
+
+Starting at the highest level and drilling down from there,
+the tasks of rendering are:
+
+1. _Render contents_ into pixels on the screen.
+1. _Animate visual effects_ on the contents from one state to another.
+1. _Scroll_ in response to input.
+1. _Route input_ efficiently to the right places so that developer scripts and other subsystems can respond.
+
+The _contents_ to render are a tree of _frames_ for each browser tab, plus the browser UI.
+And, a stream of raw input events from touch screens,
+mice, keyboards, and other hardware devices.
+
+Each frame includes:
+
+1. DOM state
+1. CSS
+1. Canvases
+1. External resources, such as images, video, fonts, and SVG
+
+A _frame_ is an HTML document, plus its URL.
+A web page loaded in a browser tab has a top-level frame,
+child frames for each iframe included in the top-level document,
+and their recursive iframe descendants.
+
+A _visual effect_ is a graphical operation applied to a bitmap,
+such as scroll, transform, clip, filter, opacity, or blend.
+
+## Architecture components
+
+In RenderingNG, these tasks are split logically across several stages and code components.
+The components end up in various CPU processes, threads,
+and sub-components within those threads.
+Each plays an important role in achieving
+[reliability](/blog/renderingng/#reliability),
+[scalable performance](/blog/renderingng/#scalable-performance)
+and [extensibility](/blog/renderingng/#extensibility-the-right-tools-for-the-job) for all web content.
+
+### Rendering pipeline structure
+
+{% Img
+src="image/ZDZVuXt6QqfXtxkpXcPGfnygYjd2/jsWIjoou2iaN6ORvMfvM.jpeg",
+alt="Diagram of the rendering pipeline as explained in the following text.",
+width="800", height="1273" %}
+
+Rendering proceeds in a pipeline with a number of stages and artifacts created along the way.
+Each _stage_ represents code that does one well-defined task within rendering.
+The _artifacts_ are data structures that are inputs or outputs of the stages;
+in the diagram inputs or outputs are indicated with arrows.
+
+This blog post won't go into much detail about the artifacts;
+that will be discussed in the next post:
+_Key Data Structures and their roles in RenderingNG_.
+
+### The pipeline stages
+
+In the preceeding diagram, stages are notated with colors indicating in which thread or process they execute:
+
+- Render process main thread: green
+- Render process compositor: yellow
+- Viz process: orange
```

Comment:
```suggestion
- **Green:** render process main thread
- **Yellow:** render process compositor
- **Orange:** viz process
```

Reply:
I think this formatting makes the text easier to process and it's visually better balanced